File: 백준/Gold/2023. 신기한 소수/신기한 소수.py
def is_prime(num):
    if num <= 1 :
        return False

    i = 2
    while i * i <= num:
        if num % i == 0:
            return False
        i += 1
    return True

def solve(cur, cnt):
    if cnt == 0:
        if is_prime(cur):
            ans.append(cur)
        return

    for i in range(10):
        nxt = cur * 10 + i
        if is_prime(nxt):
            solve(nxt, cnt - 1)

if __name__ == '__main__':
    N = int(input())
    ans = []
    solve(0,N)
    print('\n'.join(map(str, ans)))

File: 백준/Gold/2023. 신기한 소수/신기한 소수.java
import java.util.*;
import java.io.*;

public class Main {
	static int n;
	static int number[];
	static StringBuilder sb = new StringBuilder();
	
	static boolean isPrime(int num) {
		if(num <= 1)
			return false;
		
		for(int i = 2; i*i <= num;i++)
			if(num % i == 0)
				return false;
		return true;
	}
	
	
	static void solve(int cur, int cnt) {
		if(cnt == 0) {
			if(isPrime(cur))
				sb.append(cur).append('\n');
			return;
		}
		
		for(int i = 0; i<=9;i++) {
			int nxt = cur * 10 + i;
			if(isPrime(nxt))
				solve(nxt, cnt  - 1);
		}
	}
	
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		n = Integer.parseInt(br.readLine());
		number = new int[n];
		
		solve(0,n);
		System.out.println(sb);
	}
}